clang/lib/Basic/Targets/AMDGPU.cpp

```diff
@@ -188,8 +188,12 @@ bool AMDGPUTargetInfo::initFeatureMap(
 
   // TODO: Should move this logic into TargetParser
   std::string ErrorMsg;
-  if (!insertWaveSizeFeature(CPU, getTriple(), Features, ErrorMsg)) {
-    Diags.Report(diag::err_invalid_feature_combination) << ErrorMsg;
+  bool IsCombinationError;
+  if (!insertWaveSizeFeature(CPU, getTriple(), Features, ErrorMsg,
```

Comment:
It's better to make it return a `std::optional<...>`.

Reply:
I still need to return 2 values: error code and error message. Do you want std::optional<std::pair? here?

Reply:
Sure, that sounds good.

Reply:
Ugh. Actually it would need to include clang header into the TargetParser.h. This is the primary reason the diagnostics was not moved there completely and the TODO comment 3 lines above.

Reply:
Hmm, okay. Can we make the function return an integer (0 --> no error)? Or better, an enum?

Reply:
Changed. Do you like it better?